src/server/snap.rs

```diff
@@ -86,41 +85,39 @@ fn send_snap(mgr: SnapManager, addr: SocketAddr, data: ConnData) -> Result<()> {
     let snap = data.msg.get_raft().get_message().get_snapshot();
     let key = try!(SnapKey::from_snap(&snap));
     mgr.wl().register(key.clone(), SnapEntry::Sending);
-    let snap_file = box_try!(mgr.rl().get_snap_file(&key, true));
+    let mut s = box_try!(mgr.rl().get_snapshot_for_sending(&key));
     defer!({
-        snap_file.delete();
         mgr.wl().deregister(&key, &SnapEntry::Sending);
     });
-    if !snap_file.exists() {
-        return Err(box_err!("missing snap file: {:?}", snap_file.path()));
+    if !s.exists() {
+        return Err(box_err!("missing snap file: {:?}", s.path()));
     }
     // snapshot file has been validated when created, so no need to validate again.
 
-    let mut f = try!(File::open(snap_file.path()));
     let mut conn = try!(TcpStream::connect(&addr));
     try!(conn.set_nodelay(true));
     try!(conn.set_read_timeout(Some(Duration::from_secs(DEFAULT_READ_TIMEOUT))));
     try!(conn.set_write_timeout(Some(Duration::from_secs(DEFAULT_WRITE_TIMEOUT))));
 
     let res = rpc::encode_msg(&mut conn, data.msg_id, &data.msg)
-        .and_then(|_| io::copy(&mut f, &mut conn).map_err(From::from))
+        .and_then(|_| io::copy(&mut s, &mut conn).map_err(From::from))
         .and_then(|_| conn.read(&mut [0]).map_err(From::from))
         .map(|_| ())
         .map_err(From::from);
-    let size = snap_file.meta().map(|m| m.len()).unwrap_or(0);
+    let size = try!(s.total_size());
     info!("[region {}] sent snapshot {} [size: {}, dur: {:?}]",
           key.region_id,
           key,
           size,
           timer.elapsed());
-
+    s.delete();
```

Comment:
Because the io::copy use `&mut s` to read the snapshot, it would raise a "mutable refer after immutable refer" error if  `s.delete` is kept inside the defer! block. Here it just simply delete snapshot when the snapshot is done being sent. It's fine to leave the snapshot when error happens, since there is a mechanism to GC old snapshots.